### ISSUE TITLE
Method name missing from coverage report

### ISSUE THREAD

**Issue by twifty** [sebastianbergmann/php-code-coverage#605]
| Q                         | A
| --------------------------| ---------------
| php-code-coverage version | 6.2.1
| PHP version               | 7.1
| Driver                    | Xdebug / PHPDBG
| Xdebug version (if used)  | 2.6.0
| Installation Method       | Composer / PHPUnit PHAR
| Usage Method              | PHPUnit / other
| PHPUnit version (if used) | 6.2.1

I'm parsing the output coverage file in nodejs, when I discovered this. The project being tested is [symfony/framework-bundle@4.0.8](https://github.com/symfony/framework-bundle/tree/v4.0.8). One of the files causing the problem is for class `\Controller\ControllerNameParser`.

This is the relevent part of the output:
```
<line num="104" type="stmt" count="1"/>
<line num="112" type="method" name="anonymous function" complexity="1" crap="1" count="2"/>
<line num="113" type="stmt" count="2"/>
```

Compare this to the [source line 110](https://github.com/symfony/framework-bundle/blob/3571d235434b566aea39d8f8bfe38860344fd9a3/Controller/ControllerNameParser.php#L110) You will see the function `findAlternative` declared. This is missing from the coverage output, however the anonymous function on line 112 is declared in its place.

**Comment by sebastianbergmann** [comment]
Thank you for your report.

Please provide a **minimal**, **self-contained**, reproducing test case that shows the problem you are reporting.

Without such a minimal, self-contained, reproducing test case I will not be able to investigate this issue.

**Comment by twifty** [comment]
@sebastianbergmann I have no time for your arrogance. `composer create-project symfony/framework-bundle`. How much more self contained do you need?

**Comment by sebastianbergmann** [comment]
And I have no time for looking at issues that do not come with a **minimal**, reproducing example.

**Comment by sebastianbergmann** [comment]
```
$ git clone https://github.com/symfony/framework-bundle.git
```
```
$ cd framework-bundle
```
```
$ git checkout v4.0.8
```
```
$ composer info | sort
doctrine/annotations              dev-master 0f589e6 Docblock Annotations Parser
doctrine/cache                    dev-master 9f380c3 Caching library offering an object-oriented API for many cache backends
doctrine/lexer                    dev-master 0ae9b60 Base library for a lexer that can be used in Top-Down, Recursive Descent Parsers.
fig/link-util                     dev-master 1a07821 Common utility implementations for HTTP links
phpdocumentor/reflection-common   dev-master 8133918 Common reflection classes used by phpdocumentor to reflect the code structure
phpdocumentor/reflection-docblock dev-master 1826097 With this component, a library can provide support for annotations via DocBlocks or otherwise retrieve information that is embedded in a DocBlock.
phpdocumentor/type-resolver       dev-master 69bf1b1
psr/cache                         dev-master 78c5a01 Common interface for caching libraries
psr/container                     dev-master 2cc4a01 Common Container Interface (PHP FIG PSR-11)
psr/link                          dev-master 131b5c3 Common interfaces for HTTP links
psr/log                           dev-master 3490ba5 Common interface for logging libraries
psr/simple-cache                  dev-master 408d5ea Common interfaces for simple caching
symfony/asset                     dev-master 362dab9 Symfony Asset Component
symfony/browser-kit               dev-master f2a918b Symfony BrowserKit Component
symfony/cache                     dev-master ac30f70 Symfony Cache component with PSR-6, PSR-16, and tags
symfony/config                    dev-master 86b8ecd Symfony Config Component
symfony/console                   dev-master 21fe515 Symfony Console Component
symfony/css-selector              dev-master 797df60 Symfony CssSelector Component
symfony/debug                     dev-master c2c7c28 Symfony Debug Component
symfony/dependency-injection      dev-master 41ece70 Symfony DependencyInjection Component
symfony/dom-crawler               dev-master b1e0e5e Symfony DomCrawler Component
symfony/event-dispatcher          dev-master 2391ed2 Symfony EventDispatcher Component
symfony/expression-language       dev-master 2d0ffaa Symfony ExpressionLanguage Component
symfony/filesystem                dev-master 272969a Symfony Filesystem Component
symfony/finder                    dev-master d78fff5 Symfony Finder Component
symfony/form                      dev-master 92cb576 Symfony Form Component
symfony/http-foundation           dev-master 63d41b6 Symfony HttpFoundation Component
symfony/http-kernel               dev-master d911deb Symfony HttpKernel Component
symfony/inflector                 dev-master 242a9dd Symfony Inflector Component
symfony/intl                      dev-master e41f8ea A PHP replacement layer for the C intl extension that includes additional data from the ICU library.
symfony/lock                      dev-master eed6b3e Symfony Lock Component
symfony/options-resolver          dev-master a03cddf Symfony OptionsResolver Component
symfony/polyfill-intl-icu         dev-master 254919c Symfony polyfill for intl's ICU-related data and classes
symfony/polyfill-mbstring         dev-master 78be803 Symfony polyfill for the Mbstring extension
symfony/polyfill-php72            dev-master 8eca20c Symfony polyfill backporting some PHP 7.2+ features to lower PHP versions
symfony/process                   dev-master 1ed51bf Symfony Process Component
symfony/property-access           dev-master 54d1226 Symfony PropertyAccess Component
symfony/property-info             dev-master 9d8d200 Symfony Property Info Component
symfony/routing                   dev-master 04e063d Symfony Routing Component
symfony/security                  dev-master f427032 Symfony Security Component
symfony/serializer                dev-master acb8164 Symfony Serializer Component
symfony/stopwatch                 dev-master 07463bb Symfony Stopwatch Component
symfony/templating                dev-master 8a85774 Symfony Templating Component
symfony/translation               dev-master 7308ac2 Symfony Translation Component
symfony/validator                 dev-master 5adc133 Symfony Validator Component
symfony/var-dumper                dev-master 0ab3af0 Symfony mechanism for exploring and dumping PHP variables
symfony/web-link                  dev-master 6719851 Symfony WebLink Component
symfony/workflow                  dev-master 195fc22 Symfony Workflow Component
symfony/yaml                      dev-master abab82a Symfony Yaml Component
twig/twig                         2.x-dev b4758ab    Twig, the flexible, fast, and secure template language for PHP
webmozart/assert                  dev-master 0df1908 Assertions to validate method input/output with nice error messages.
```
```
$ phpunit --coverage-clover clover.xml --verbose
PHPUnit 7.1.2 by Sebastian Bergmann and contributors.

Runtime:       PHP 7.2.4 with Xdebug 2.6.0
Configuration: /home/sb/framework-bundle/phpunit.xml.dist

.............................................................   61 / 1061 (  5%)
.............................................................  122 / 1061 ( 11%)
.............................................................  183 / 1061 ( 17%)
.E...........................................................  244 / 1061 ( 22%)
.......................E.....................................  305 / 1061 ( 28%)
.............................................................  366 / 1061 ( 34%)
.............................................................  427 / 1061 ( 40%)
...........S.................................................  488 / 1061 ( 45%)
.............................................................  549 / 1061 ( 51%)
.............................................................  610 / 1061 ( 57%)
......SS.....................................................  671 / 1061 ( 63%)
.............................................................  732 / 1061 ( 68%)
..........SSSS...............................................  793 / 1061 ( 74%)
.............................................................  854 / 1061 ( 80%)
........................................SSSS.................  915 / 1061 ( 86%)
.............................................................  976 / 1061 ( 91%)
............................................................. 1037 / 1061 ( 97%)
........................                                      1061 / 1061 (100%)

Time: 4.59 minutes, Memory: 162.00MB

There were 2 errors:

1) Symfony\Bundle\FrameworkBundle\Tests\Controller\AbstractControllerTest::testAddFlash
PHPUnit\Framework\Exception: PHP Fatal error:  Interface 'Symfony\Bundle\FrameworkBundle\Tests\Fixtures\Validation\NotExistingInterface' not found in /home/sb/framework-bundle/Tests/Fixtures/Validation/Article.php on line 5
PHP Stack trace:
PHP   1. {main}() Standard input code:0
PHP   2. require_once() Standard input code:659

Fatal error: Interface 'Symfony\Bundle\FrameworkBundle\Tests\Fixtures\Validation\NotExistingInterface' not found in /home/sb/framework-bundle/Tests/Fixtures/Validation/Article.php on line 5

Call Stack:
    0.0009     787840   1. {main}() Standard input code:0
    0.0345   10380688   2. require_once('/home/sb/framework-bundle/Tests/Fixtures/Validation/Article.php') Standard input code:659

2) Symfony\Bundle\FrameworkBundle\Tests\Controller\ControllerTest::testAddFlash
PHPUnit\Framework\Exception: PHP Fatal error:  Interface 'Symfony\Bundle\FrameworkBundle\Tests\Fixtures\Validation\NotExistingInterface' not found in /home/sb/framework-bundle/Tests/Fixtures/Validation/Article.php on line 5
PHP Stack trace:
PHP   1. {main}() Standard input code:0
PHP   2. require_once() Standard input code:659

Fatal error: Interface 'Symfony\Bundle\FrameworkBundle\Tests\Fixtures\Validation\NotExistingInterface' not found in /home/sb/framework-bundle/Tests/Fixtures/Validation/Article.php on line 5

Call Stack:
    0.0010     794208   1. {main}() Standard input code:0
    0.0344   10387056   2. require_once('/home/sb/framework-bundle/Tests/Fixtures/Validation/Article.php') Standard input code:659

--

There were 11 skipped tests:

1) Symfony\Bundle\FrameworkBundle\Tests\DependencyInjection\XmlFrameworkExtensionTest::testAssetsHelperIsRemovedWhenPhpTemplatingEngineIsEnabledAndAssetsAreDisabled
The assets key cannot be set to false using the XML configuration format.

/home/sb/framework-bundle/Tests/DependencyInjection/XmlFrameworkExtensionTest.php:28

2) Symfony\Bundle\FrameworkBundle\Tests\Functional\CachePoolsTest::testRedisCachePools
Extension redis is required.

3) Symfony\Bundle\FrameworkBundle\Tests\Functional\CachePoolsTest::testRedisCustomCachePools
Extension redis is required.

4) Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperDivLayoutTest::testHelp
Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperDivLayoutTest::renderHelp() is not implemented.

/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:119
/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:423

5) Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperDivLayoutTest::testHelpNotSet
Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperDivLayoutTest::renderHelp() is not implemented.

/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:119
/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:438

6) Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperDivLayoutTest::testHelpSetLinkFromWidget
Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperDivLayoutTest::renderHelp() is not implemented.

/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:119
/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:452

7) Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperDivLayoutTest::testHelpNotSetNotLinkedFromWidget
Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperDivLayoutTest::renderHelp() is not implemented.

/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:119
/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:468

8) Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperTableLayoutTest::testHelp
Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperTableLayoutTest::renderHelp() is not implemented.

/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:119
/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:423

9) Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperTableLayoutTest::testHelpNotSet
Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperTableLayoutTest::renderHelp() is not implemented.

/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:119
/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:438

10) Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperTableLayoutTest::testHelpSetLinkFromWidget
Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperTableLayoutTest::renderHelp() is not implemented.

/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:119
/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:452

11) Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperTableLayoutTest::testHelpNotSetNotLinkedFromWidget
Symfony\Bundle\FrameworkBundle\Tests\Templating\Helper\FormHelperTableLayoutTest::renderHelp() is not implemented.

/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:119
/home/sb/framework-bundle/vendor/symfony/form/Tests/AbstractLayoutTest.php:468

ERRORS!
Tests: 1061, Assertions: 2178, Errors: 2, Skipped: 11.

Generating code coverage report in Clover XML format ... done
```
```
$ sed -n 110,132p Controller/ControllerNameParser.php 
    private function findAlternative(string $nonExistentBundleName): ?string
    {
        $bundleNames = array_map(function ($b) {
            return $b->getName();
        }, $this->kernel->getBundles());

        $alternative = null;
        $shortest = null;
        foreach ($bundleNames as $bundleName) {
            // if there's a partial match, return it immediately
            if (false !== strpos($bundleName, $nonExistentBundleName)) {
                return $bundleName;
            }

            $lev = levenshtein($nonExistentBundleName, $bundleName);
            if ($lev <= strlen($nonExistentBundleName) / 3 && (null === $alternative || $lev < $shortest)) {
                $alternative = $bundleName;
                $shortest = $lev;
            }
        }

        return $alternative;
    }
```
`findAlternative()` (declared on line 110) does not show up in the Clover XML (see below).
```
$ sed -n 2679,2681p clover.xml
        <line num="101" type="stmt" count="1"/>
        <line num="104" type="stmt" count="1"/>
        <line num="112" type="method" name="anonymousFunction:112#547" complexity="1" crap="1" count="2"/>
```
TL;DR: After executing 465807 LOC I am able to reproduce your issue.

**Comment by sebastianbergmann** [comment]
`SebastianBergmann\CodeCoverage\Report\Clover` has the following code:
```php
foreach ($report as $item) {
    // ...

    foreach ($classes as $className => $class) {
        // ...

        foreach ($class['methods'] as $methodName => $method) {
            if ($method['executableLines'] == 0) {
                continue;
            }

            // ...
```
For `findAlternative()`, the `$method['executableLines'] == 0` expression evaluates to `true`. This is why `findAlternative()` is missing from the Clover XML logfile.

Instead of one problem this seems to be two problems:

* Why is `$method['executableLines']` (which it should not be)?
* Why are methods that do not have executable lines ignored?

**Comment by twifty** [comment]
Are these 'anonymous function' method names in the clover report also in error? Since, strictly speaking they are functions and not class methods.

**Comment by sebastianbergmann** [comment]
To me, anonymous functions are implementation details of the scope that declare and use them. In this case that would be the `findAlternative()` method.

Details for anonymous functions are only included in the code coverage report(s) because that was requested on multiple occasions in the past.

TL;DR: As it stands, `anonymousFunction:112#547` is not included in error but intentionally.

**Comment by twifty** [comment]
Ok. This means merging two clover reports is next to impossible. Since, with the use of 'anonymous function' we have no way of knowing which lines belong to the inner scope of the function and which lines belong to the outer method.

This comes from trying calculate the covered methods from the merged lines, and hence the covered elements. My only recourse is to compare the line numbers against the parsed file, adding a layer of complexity with async requests for each file in the report.
